Extract logic of printing out to console

jingweno opened the pull request:
Introduce a `console` package which encapsulates logic of printing out
to console. This provides flexibility of discarding output in tests.
It also allows us to unify format of printing to console (if we want to).

This fixes #801.

jingweno commented on commands/commands_test.go:
```diff
@@ -1,11 +1,16 @@
 )
 
+func init() {
+	UI = ui{Stdout: ioutil.Discard, Stderr: ioutil.Discard}
```
This is the code muting the print out to console in tests

jingweno commented on commands/commands_test.go:
```diff
@@ -1,11 +1,19 @@
 )
 
+func TestMain(m *testing.M) {
+	console.Default = console.Console{Stdout: ioutil.Discard, Stderr: ioutil.Discard}
```
I did some refactoring again. This is the new code to mute the print out to console in tests

mislav replied:
Since TestMain affects all tests everywhere, should we put it in a file where it's more apparent that this is a global setting?

Also, shouldn't we capture the output instead of discard it?

jingweno replied:
> Since TestMain affects all tests everywhere

This `TestMain` will only affect the commands package. We'd need to put it into a test file in the rest of packages if we want this to be global.

> Also, shouldn't we capture the output instead of discard it?

I'm not aware of any test that asserts the output yet. For [`TestCommandCall`](https://github.com/github/hub/blob/master/commands/commands_test.go#L130) mentioned in #801, we're checking the the function is called and the arguments are correctly passed in.

mislav commented:
It's a good start, but "console" might be a slight misnomer because a unix executable doesn't know and shouldn't care where its output is going or who's going to read it.

Is there an option to keep our `fmt.Println` and similar but somehow capture and redirect all standard output in tests to a string so we can inspect it? This is trivial in Ruby.

jingweno commented:
> "console" might be a slight misnomer because a unix executable doesn't know and shouldn't care where its output is going

Any suggestion? I'm thinking something as generic as `ui.Infoln`. or `log.Infoln`.

> Is there an option to keep our fmt.Println and similar but somehow capture and redirect all standard output in tests to a string so we can inspect it?

Yup, `Console` expects `Stdout` and `Stderr` to be `io.Writer`, which is an interface to all writer, including [bytes of string](http://golang.org/pkg/bytes/#Buffer), so:

```go
var b bytes.Buffer
console.Default = Console{Stdout: &b, Stderr: &b}
// do something with b
```

jingweno commented:
@mislav Any more feedback here?

mislav commented:
I think "UI" is a good name, because it's an interface with whoever is using the tool from the outside. I'd also like to keep the `Println` and `Printf` names.

How about:

* `ui.Println`
* `ui.Printf`
* `ui.Errorln`
* `ui.Errorf`

jingweno commented:
> I think "UI" is a good name, because it's an interface with whoever is using the tool from the outside.

I've extracted `UI` into [an interface](https://github.com/github/hub/blob/fix_test_output/ui/ui.go) and `Console` as an implementation.

> I'd also like to keep the Println and Printf names.

After a second thought, I agree `Println` and `Printf` are better since they are the same as the ones from `fmt`. Fixed with `--amend`.

jingweno commented:
Merging since all comments are addressed.
